bemani/backend/popn/unilab.py

```diff
@@ -0,0 +1,640 @@
+                riddle = math.floor(random.randrange(1, 21, 1))
+                try:
+                    randomRiddles.index(riddle)
+                except ValueError:
+                    break
+
+            randomRiddles.append(riddle)
+
+            sh_riddles = Node.void("sh_riddles")
+            riddles_data.add_child(sh_riddles)
+            sh_riddles.add_child(Node.u32("sh_riddles_id", riddle))
+
+        # Set up kaimei riddles achievements
+        achievements = self.data.local.user.get_achievements(self.game, self.version, userid)
+        for achievement in achievements:
+            if achievement.type == "riddle":
+                kaimei_gauge = achievement.data.get_int("kaimei_gauge")
+                is_cleared = achievement.data.get_bool("is_cleared")
+                riddles_cleared = achievement.data.get_bool("riddles_cleared")
+                select_count = achievement.data.get_int("select_count")
+                other_count = achievement.data.get_int("other_count")
+
+                sp_riddles = Node.void("sp_riddles")
+                riddles_data.add_child(sp_riddles)
+                sp_riddles.add_child(Node.u16("kaimei_gauge", kaimei_gauge))
+                sp_riddles.add_child(Node.bool("is_cleared", is_cleared))
+                sp_riddles.add_child(Node.bool("riddles_cleared", riddles_cleared))
+                sp_riddles.add_child(Node.u8("select_count", select_count))
+                sp_riddles.add_child(Node.u32("other_count", other_count))
+
+        # Narunaru♪ UniLab jikkenshitsu! event
+        event_p27 = Node.void("event_p27")
+        root.add_child(event_p27)
+        event_p27.add_child(Node.s16("team_id", profile.get_int("team_id")))
+        event_p27.add_child(Node.bool("first_play", profile.get_bool("first_play", True)))
+        event_p27.add_child(Node.s16("select_battery_id", profile.get_int("select_battery_id", 1)))
+        event_p27.add_child(Node.bool("elem_first_play", profile.get_bool("elem_first_play", True)))
+        event_p27.add_child(Node.bool("today_first_play", profile.get_bool("today_first_play", True)))
+
+        # Set up Narunaru♪ UniLab jikkenshitsu! achievements
+        for achievement in achievements:
+            if achievement.type == "lab":
+                team_id = achievement.data.get_int("team_id")
+                ex_no = achievement.data.get_int("ex_no")
+                point = achievement.data.get_int("point")
+                is_cleared = achievement.data.get_bool("is_cleared")
+
+                team = Node.void("team")
+                event_p27.add_child(team)
+                team.add_child(Node.s16("team_id", team_id))
+                team.add_child(Node.s16("ex_no", ex_no))
+                team.add_child(Node.u32("point", point))
+                team.add_child(Node.bool("is_cleared", is_cleared))
+
+        # Set up Kakusei no Elem achievements
+        game_config = self.get_game_config()
+        if game_config.get_bool("force_unlock_deco"):
```

Comment:
We are setting maximum energy and battery here, but in the `unformat_profile` function we don't ignore updating this value if `force_unlock_deco` is true, which will poison user profiles with fully unlocked events.